// app/constants/formulas.js
import * as properties from "app/constants/properties";

export function difficulty(linesCleared: number): number {
    if (linesCleared === 0) return 0;
    return Math.min(properties.MAX_DIFFICULTY, Math.floor(Math.log(linesCleared) / properties.DIFFICULTY_LOG_BASE));
}

export function difficultyLines(difficulty: number): number {
    if (difficulty === 0) return 0;
    return Math.pow(properties.DIFFICULTY_BASE, difficulty);
}

export function delay(difficulty: number): number {
    return properties.DELAY - difficulty * properties.DELAY_AMOUNT;
}

export function score(rows: number, difficulty: number, combo: number): number {
    return properties.ROWS_SCORE[rows - 1] * (combo + 1) * (difficulty + 1);
}

export function level(experience: number) {
    return Math.min(properties.MAX_LEVEL, Math.floor(experience === 0 ? 0 : Math.log(experience) / properties.LEVEL_LOG_BASE));
}

export function levelExperience(level: number): number {
    if (level === 0) return 0;
    return Math.pow(properties.LEVEL_BASE, level);
}
