// app/components/Playground/Playground.js
// @flow

import {observer} from "mobx-react";
import React from "react";

import type {Position} from "app/models/Matrix";
import Matrix from "app/models/Matrix";
import Piece from "app/models/Piece";
import Effect from "app/models/Effect";
import Block from "app/models/Block";

import {ROWS} from "app/constants/properties";

import PlaygroundGrid from "app/components/PlaygroundGrid/PlaygroundGrid";
import MatrixComponent from "app/components/Matrix/Matrix";
import Effects from "app/components/Effects/Effects";

import "./Playground.scss";

@observer
export default class Playground extends React.PureComponent {
    propsTypes: {
        matrix: Matrix<Block>,
        piece: Piece,
        piecePosition: Position,
        effects: Effect[]
    };

    render() {
        const blockSize = Math.min((window.innerHeight - 200) / ROWS, 32);

        const rows = this.props.matrix.rows();
        const cols = this.props.matrix.cols();

        const pieceMatrixLeft = this.props.piecePosition.x * blockSize + 'px';
        const pieceMatrixTop = this.props.piecePosition.y * blockSize + 'px';

        return (
            <div className="Playground" style={{width: `${cols * blockSize}px`, height: `${rows * blockSize}px`}}>
                <div className="Playground-background"/>
                <PlaygroundGrid rows={rows} cols={cols}/>

                <MatrixComponent matrix={this.props.matrix} blockSize={blockSize}/>
                <MatrixComponent key={this.props.piece.id} matrix={this.props.piece} blockSize={blockSize}
                                 style={{left: pieceMatrixLeft, top: pieceMatrixTop}}/>

                <Effects effects={this.props.effects} blockSize={blockSize}/>
            </div>
        )
    }
}
