// app/components/Playground/Playground.js
// @flow

import {observer} from "mobx-react";
import React from "react";

import type {Position} from "app/models/Matrix";
import Matrix from "app/models/Matrix";
import Piece from "app/models/Piece";
import Effect from "app/models/Effect";
import Block from "app/models/Block";

import {COLS, ROWS} from "app/constants/properties";

import PlaygroundGrid from "app/components/PlaygroundGrid/PlaygroundGrid";
import MatrixComponent from "app/components/Matrix/Matrix";
import Effects from "app/components/Effects/Effects";

import "./Playground.scss";

@observer
export default class Playground extends React.PureComponent {
    propsTypes: {
        matrix: Matrix<Block>,
        piece: Piece,
        piecePosition: Position,
        effects: Effect[]
    };

    render() {
        let blockSize;

        if (window.innerWidth <= 767) {
            const blockHeight = Math.min((window.innerHeight - 90) / ROWS, 32);
            const blockWidth = Math.min((window.innerWidth - 20) / COLS, 32);
            blockSize = Math.min(blockHeight, blockWidth);
        } else {
            blockSize = Math.min((window.innerHeight - 200) / ROWS, 32);
        }

        const rows = this.props.matrix.rows();
        const cols = this.props.matrix.cols();

        const pieceMatrixLeft = this.props.piecePosition.x * blockSize + 'px';
        const pieceMatrixTop = this.props.piecePosition.y * blockSize + 'px';

        return (
            <div className="Playground" style={{width: `${cols * blockSize}px`, height: `${rows * blockSize}px`}}>
                <div className="Playground-background"/>
                <PlaygroundGrid rows={rows} cols={cols}/>

                <MatrixComponent matrix={this.props.matrix} blockSize={blockSize}/>
                <div className="Playground-matrix-container" style={{left: pieceMatrixLeft, top: pieceMatrixTop}}>
                    <MatrixComponent key={this.props.piece.id} matrix={this.props.piece} blockSize={blockSize}/>
                </div>

                <Effects effects={this.props.effects} blockSize={blockSize}/>
            </div>
        )
    }
}
