// app/components/Spell/Spell.js
// @flow

import React from "react";
import SpellModel from "app/models/Spell";
import "./Spell.scss";

export default class Spell extends React.PureComponent {
    props: {
        spell: SpellModel,
        castKey?: string
    };

    render() {
        const spell = this.props.spell.constructor;

        return (
            <div className="Spell">
                <button type="button" className="Spell-icon">
                    <img src={spell.icon} alt={`${spell.title} spell icon`}/>
                </button>

                {this.renderCastKey()}

                <div className="Spell-popover">
                    <div className="Spell-title">{spell.title}</div>
                    <div className="Spell-description">{spell.title}</div>
                </div>
            </div>
        )
    }

    renderCastKey() {
        const castKey = this.props.castKey;
        if (!castKey) return null;

        return (
            <div className="Spell-key">{castKey}</div>
        );
    }
};
