// app/components/Background/Background.js
// @flow

import React from "react";
import "./Background.scss";

export default class Background extends React.PureComponent {
    props: {
        image: string,
        mask?: any,
        maskColor?: any,
        size?: any,
        blur?: any,
        repeat?: boolean
    };

    render() {
        const mask = this.props.mask || .65;
        const maskColor = this.props.maskColor || 'black';
        const size = this.props.size || 'cover';
        const blur = this.props.blur || false;
        const repeat = this.props.repeat || 'no-repeat';
        const position = this.props.position || 'center top';

        const imageStyle = {
            backgroundImage: `url('${this.props.image}')`,
            backgroundSize: size,
            backgroundRepeat: repeat,
            backgroundPosition: position
        };

        const maskStyle = {
            background: maskColor,
            opacity: mask
        };

        return (
            <div className={`Background ${blur ? 'Background--blurred' : ''}`}>
                <div className="Background-image" style={imageStyle}/>
                <div className="Background-mask" style={maskStyle}/>
            </div>
        )
    }
}
