// app/stores/GameStore.js
// @flow

import {action, computed, observable} from "mobx";
import Store from "app/stores/Store";

import Character from "app/models/Character";
import Piece from "app/models/Piece";
import Spell from "app/models/Spell";
import Block from "app/models/Block";
import type {Position} from "app/models/Matrix";
import Matrix from "app/models/Matrix";
import Effect from "app/models/Effect";
import Stage from "app/models/Stage";
import TimePlayer from "app/models/TimePlayer";

import * as gameStates from "app/constants/gameStates";
import * as gameProperties from "app/constants/gameProperties";
import * as gameModes from "app/constants/gameModes";

import MoveEvent from "app/game/events/MoveEvent";
import CastEvent from "app/game/events/CastEvent";

import ControlManager from "app/controls/ControlManager";
import KeyboardControl from "app/controls/KeyboardControl";

class GameStore extends Store {
    characterClass: Function;
    @observable character: Character;

    stageClass: Function;
    @observable stage: Stage;
    @observable mode: string = gameModes.SURVIVAL;

    controlManager: ControlManager;

    @observable piece: Piece;

    @observable state: string;
    @observable speed: number = 0;

    @observable piecePosition: Position;
    @observable matrix: Matrix<Block>;
    @observable effects: Effect[] = [];

    timer: TimePlayer;

    constructor() {
        super();

        this.controlManager = new ControlManager({
            move: this.move,
            drop: this.drop,
            rotate: this.rotate,
            cast: this.cast,
            pause: this.pause,
            resume: this.resume
        }, [KeyboardControl]);
    }

    /*
     * Menu actions
     */
    @action.bound selectCharacter(characterClass: Function) {
        this.characterClass = characterClass;
    }

    @action.bound selectStage(stageClass: Function) {
        this.stageClass = stageClass;
    }

    /*
     * Control driven actions
     */

    @action.bound move(x: number, y: number): boolean {
        if (!this.playing) return false;

        if (x > 0 && y > 0) y = 0;

        let newPosition = {x: this.piecePosition.x + x, y: this.piecePosition.y + y};
        let canMove = this.matrix.gridAvailable(this.piece.grid, newPosition);

        if (!canMove) {
            if (y > 0) {
                if (this.piecePosition.y < 0) {
                    this.gameOver();
                } else {
                    this.newPiece();
                }
            }

            return false;
        }

        if (y > 0) {
            this.timer.reset();
        }

        let event = new MoveEvent(this.piecePosition, newPosition);
        if (!event.fireBefore()) return false;

        this.piecePosition = newPosition;
        event.fire();

        return true;
    }

    @action.bound drop(): void {
        const y = this.matrix.landingGridPosition(this.piece.grid, this.piecePosition).y - this.piecePosition.y;
        this.move(0, y);
        this.decay();
        this.newPiece();
    }

    @action.bound rotate(): void {
        if (!this.playing || !this.matrix.gridAvailable(this.piece.rotated(), this.piecePosition)) return;

        this.piece.rotate();
        // let event = new RotateEvent();
    }

    @action cast(spell: Spell): void {
        if (!spell.canCast()) return;

        let event = new CastEvent(spell);
        if (!event.fireBefore()) return;

        spell.cast();
        event.fire();
    }

    @action.bound resume(): void {
        if (!this.paused) return;

        this.state = gameStates.PLAYING;
        this.timer.play();
    }

    @action.bound pause(): void {
        if (!this.playing) return;

        this.timer.pause();
        this.state = gameStates.PAUSED;
    }

    /*
     * State changing actions
     */

    @action.bound play(): void {
        this.matrix = new Matrix(gameProperties.ROWS, gameProperties.COLS);

        this.state = gameStates.PLAYING;

        this.character = new this.characterClass();
        this.stage = new this.stageClass();
        this.newPiece();

        this.timer = new TimePlayer(gameProperties.SPEED, () => this.decay());
        this.timer.play();
    }

    @action.bound gameOver(): void {
        this.state = gameStates.LOSS;
        this.effects = [];
        this.timer.pause();
        alert('components Over');
    }

    /*
     * Internal actions and actions available for spells
     */

    @action.bound newPiece(piece: Piece | null = null, piecePosition: Position | null = null): void {
        if (!this.playing) return;

        piece = piece || Piece.random(this.character);
        if (this.piece) this.matrix.apply(this.piece.grid, this.piecePosition);

        if (!piecePosition) piecePosition = {
            x: Math.floor(Math.random() * (this.matrix.cols() - piece.cols())),
            y: -piece.rows()
        };

        this.piece = piece;
        this.piecePosition = piecePosition;

        this.matrix.clearFullRows();
    }

    @action.bound decay(): boolean {
        return this.move(0, 1);
    }

    @action.bound addEffect(effect: Effect, duration: number = 0): void {
        this.effects.push(effect);

        if (duration) {
            setTimeout(() => {
                const index = this.effects.indexOf(effect);
                if (index !== -1) this.effects.splice(index, 1);
            }, duration);
        }
    }

    /*
     * State shortcuts
     */

    @computed get playing(): boolean {
        return this.state === gameStates.PLAYING;
    }

    @computed get paused(): boolean {
        return this.state === gameStates.PAUSED;
    }
}

export default new GameStore();
