// app/components/Effects/Effects.js
// @flow

import {observer} from "mobx-react";
import React from "react";
import Effect from "app/models/Effect";
import "./Effects.scss";

@observer
export default class Effects extends React.PureComponent {
    props: {
        effects: Effect[],
        blockSize: number
    };

    render() {
        const blockSize = this.props.blockSize;

        const effects = this.props.effects.map((effect) => {
            const style = !effect.position ? null : {
                left: effect.position.x * blockSize + .5 + 'px',
                top: effect.position.y * blockSize + .5 + 'px'
            };

            const classes = ['Effects-effect'];
            if(effect.position) classes.push('Effects-effect--positioned');

            return (
                <div className={classes.join(' ')} key={effect.id} style={style}>
                    {effect.render(blockSize)}
                </div>
            )
        });

        return (
            <div className="effects">
                {effects}
            </div>
        );
    }
}
