// app/controls/TouchControl.js
// @flow

import Control from './Control'

// export default class TouchControl extends Control {
//     load() {
//         this.threshold = 150
//         this.tetris = this.state.tetris
//
//         this.touching = false
//         this.moved = false
//         this.decayed = false
//
//         this.state.game.input.onTap.add(this.onTap.bind(this))
//
//         let tetrisComponent = this.state.components.tetris
//         tetrisComponent.onChildInputDown.add(this.onDown.bind(this))
//         tetrisComponent.onChildInputUp.add(this.onUp.bind(this))
//     }
//
//     onTap(_, double) {
//         this.move(double)
//     }
//
//     onDown() {
//         this.reset()
//         this.moved = false
//         this.touching = true
//     }
//
//     onUp() {
//         this.touching = false
//     }
//
//     move(double = false) {
//         let map = this.state.tetris.map
//         let piece = map.piece
//         let origin = map.origin
//
//         if (!piece || !origin) return
//
//         let pointer = this.state.game.input.pointer1
//         let x = pointer.x - this.state.components.tetris.x - TetrisComponent.borderSize
//         let y = pointer.y - this.state.components.tetris.y - TetrisComponent.borderSize
//
//         let pos = {
//             x0: BlockComponent.size * origin.x,
//             y0: BlockComponent.size * origin.y,
//             x1: BlockComponent.size * (origin.x + piece.width()),
//             y1: BlockComponent.size * (origin.y + piece.height())
//         }
//
//         // This variable is here to avoid miss dropping
//         let decayed = false
//         if (x < pos.x0) {
//             this.tetris.move(-1)
//             this.moved = true
//         } else if (x > pos.x1) {
//             this.tetris.move(1)
//             this.moved = true
//         } else if (x >= pos.x0 && x <= pos.x1) {
//             if (!this.moved && y <= (pos.y1 + BlockComponent.size)) {
//                 this.tetris.rotate()
//             } else if (y > (pos.y1 + BlockComponent.size * 2)) {
//                 let action = double && this.decayed ? 'drop' : 'decay'
//                 this.tetris[action]()
//                 this.moved = true
//
//                 decayed = action == 'decay'
//             }
//         }
//
//         this.decayed = decayed
//     }
//
//     onLoop() {
//         if (this.touching && this.canMove()) {
//             this.move()
//         }
//     }
// }
