// app/game/components/Matrix/Matrix.js
import {observer} from "mobx-react";
import React from "react";

import GameMatrix from "app/game/Matrix";

import Block from "app/game/components/Block/Block";

import "./Matrix.scss";

@observer
export default class Matrix extends React.PureComponent {
    props: {
        matrix: GameMatrix,
        blockSize: number,
        style?: Map
    };

    render() {
        const blockSize = this.props.blockSize;

        const width = this.props.matrix.cols() * blockSize;
        const height = this.props.matrix.rows() * blockSize;

        const blocks = this.props.matrix.mapValue((block, x, y) => {
            const left = x * blockSize + 'px';
            const top = y * blockSize + 'px';

            return (
                <Block key={block.id} style={{left, top}} block={block} blockSize={blockSize}/>
            )
        });

        return (
            <div className="Matrix"
                 style={{width, height, ...this.props.style || {}}}>
                {blocks}
            </div>
        );
    }
}
