// app/game/Piece.js
// @flow

import Block from "app/game/Block";
import Matrix from "app/game/Matrix";

import pieceTypes from "app/config/pieceTypes";
import pieceColors from "app/config/pieceColors";

export default class Piece extends Matrix<Block> {
    type: string;

    id: number;

    static _id = 1;

    constructor(type: string, character: any = null) {
        const typeMatrix = pieceTypes[type];

        super(Matrix.rows(typeMatrix), Matrix.cols(typeMatrix));
        this.type = type;

        const color = this.color();
        Matrix.eachValue(typeMatrix, (_, x, y) => this.set({x, y}, new Block(this, color, character)));

        this.id = this.constructor._id++;
    }

    color(): any {
        return pieceColors[this.typeIndex()];
    }

    typeIndex(): number {
        return Object.keys(pieceTypes).indexOf(this.type);
    }

    detachBlocks() {
        this.eachValue((block: Block) => block.detach());
    }

    static randomType(): string {
        const index = Math.floor(Math.random() * Object.keys(pieceTypes).length);
        return Object.keys(pieceTypes)[index];
    }

    static random(character: any = null): this {
        return new this(this.randomType(), character);
    }
}
