// app/game/Piece.js
// @flow

import Block from "app/game/Block";
import Matrix from "app/game/Matrix";

import pieceTypes from "app/config/pieceTypes";
import pieceColors from "app/config/pieceColors";

export default class Piece extends Matrix<Block> {
    type: string;

    constructor(type: string) {
        const typeMatrix = pieceTypes[type];

        super(Matrix.rows(typeMatrix), Matrix.cols(typeMatrix));
        this.type = type;

        const color = this.color();
        Matrix.eachValue(typeMatrix, (_, x, y) => this.set({x, y}, new Block(this, color)));
    }

    color(): any {
        return pieceColors[this.typeIndex()];
    }

    typeIndex(): number {
        return Object.keys(pieceTypes).indexOf(this.type);
    }

    detachBlocks() {
        this.eachValue((block: Block) => block.detach());
    }

    static randomType(): number {
        return Math.floor(Math.random() * Object.keys(pieceTypes).length);
    }

    static random(): this {
        return new this(Object.keys(pieceTypes)[this.randomType()]);
    }
}
