// app/game/components/CharacterBar/CharacterBar.js
import {observer} from "mobx-react";
import React from "react";
import FontAwesome from "react-fontawesome";

import "./CharacterBar.scss";

@observer
export default class CharacterMenu extends React.PureComponent {
    props: {
        title: string,
        icon: string,
        value: number,
        max: number,
        color: any
    };

    render() {
        const percentage = 100 - this.props.value * 100 / this.props.max;
        const style = {
            clipPath: `inset(${percentage}% 0 0 0)`,
            color: this.props.color
        };

        return (
            <div className="CharacterBar">
                <div className="CharacterBar-icon">
                    <FontAwesome name={this.props.icon} className="CharacterBar-background"/>
                    <FontAwesome name={this.props.icon} className="CharacterBar-fill" style={style}/>
                </div>

                <div className="CharacterBar-text">
                    <div className="CharacterBar-title">{this.props.title}</div>
                    <div className="CharacterBar-value">{this.props.value} / {this.props.max}</div>
                </div>
            </div>
        )
    }
}
