// app/components/Matrix/Matrix.js
// @flow

import {observer} from "mobx-react";
import React from "react";
import MatrixModel from "app/models/Matrix";
import BlockModel from "app/models/Block";
import Block from "app/components/Block/Block";
import "./Matrix.scss";

@observer
export default class Matrix extends React.PureComponent {
    props: {
        matrix: MatrixModel<BlockModel>,
        blockSize: number,
        style?: { [string]: string }
    };

    render() {
        const blockSize = this.props.blockSize;

        const width = this.props.matrix.cols() * blockSize;
        const height = this.props.matrix.rows() * blockSize;

        const blocks = this.props.matrix.mapValue((block: BlockModel, {x, y}) => {
            return (
                <Block key={block.id} block={block} blockSize={blockSize} x={x} y={y}/>
            )
        });

        return (
            <div className="Matrix"
                 style={{width, height, ...this.props.style || {}}}>
                {blocks}
            </div>
        );
    }
}
