// app/components/Matrix/Matrix.js
// @flow

import {observer} from "mobx-react";
import React from "react";
import MatrixModel from "app/models/Matrix";
import BlockModel from "app/models/Block";
import Block from "app/components/Block/Block";
import "./Matrix.scss";

@observer
export default class Matrix extends React.PureComponent {
    props: {
        matrix: MatrixModel<BlockModel>,
        blockSize: number
    };

    render() {
        const blockSize = this.props.blockSize;

        const style = {
            width: this.props.matrix.cols() * blockSize,
            height: this.props.matrix.rows() * blockSize
        };

        const blocks = this.props.matrix.mapValue((block: BlockModel, {x, y}) => {
            const left = x * blockSize;
            const top = y * blockSize;

            return (
                <div key={block.id} className="Matrix-block-wrapper" style={{top, left}}>
                    <Block block={block} blockSize={blockSize} x={x} y={y}/>
                </div>
            )
        });

        return (
            <div className="Matrix" style={style}>
                {blocks}
            </div>
        );
    }
}
